libp2p/stream_muxer/mplex/mplex_stream.py

```diff
@@ -34,13 +36,42 @@ def __init__(self, stream_id: int, initiator: bool, mplex_conn: IMuxedConn) -> N
         self.local_closed = False
         self.remote_closed = False
         self.stream_lock = asyncio.Lock()
-
-    async def read(self) -> bytes:
-        """
-        read messages associated with stream from buffer til end of file
-        :return: bytes of input
-        """
-        return await self.mplex_conn.read_buffer(self.stream_id)
+        self._buf = bytearray()
+
+    async def read(self, n: int = -1) -> bytes:
+        """
+        Read up to n bytes. Read possibly returns fewer than `n` bytes,
+        if there are not enough bytes in the Mplex buffer.
+        If `n == -1`, read until EOF.
+        :param n: number of bytes to read
+        :return: bytes actually read
+        """
+        if n < 0 and n != -1:
+            raise ValueError("`n` can only be -1 if it is negative")
+        # If the buffer is empty at first, blocking wait for data.
+        if len(self._buf) == 0:
+            self._buf.extend(await self.mplex_conn.read_buffer(self.stream_id))
+        # Sanity check: `self._buf` should never be empty here.
+        if self._buf is None or len(self._buf) == 0:
+            raise Exception("`self._buf` should never be empty here")
```

Comment:
This error doesn't seem to be handled by the functions in the call stack and it should probably be noted in the docstring.